### PR TITLE
simplify sigatom code

### PR DESCRIPTION
**Pull request by vtjnash**
this doesn't fix 161, it just moves the failure condition so that it
should be much less common and more avoidable

### CODE REVIEW

File: src/GLib/signals.jl [hunk 1]
```diff
@@ -146,39 +146,39 @@ function signal_emit(w::GObject, sig::AbstractStringLike, RT::Type, args...)
...
 gtk_stack = nothing # need to call g_loop_run from only one stack
-gtk_work = nothing
-g_sigatom_flag = false
-function g_sigatom(f::Base.Callable)
-    # don't yield() from f inside g_sigatom, use g_siginterruptible(yield) if unsure
-    global g_sigatom_flag, gtk_stack, gtk_work
+gtk_yielded = false # when true, use the `gtk_doatomic` queue to run sigatom functions
+gtk_doatomic = [] # (work, notification) scheduler queue
+g_sigatom_flag = false # keep track of Base sigatomic state
+function g_sigatom(f::Base.Callable) # calls f, were f never throws (but this function may throw)
```
Comment (timholy) [inline src/GLib/signals.jl]: were->where

File: src/GLib/signals.jl [hunk 2]
```diff
@@ -146,39 +146,39 @@ function signal_emit(w::GObject, sig::AbstractStringLike, RT::Type, args...)
...
+        push!(gtk_doatomic, (f, ct))
+        return wait()
+    end
+
     if !prev
         sigatomic_begin()
         g_sigatom_flag = true
     end
-    local ret
+    local ret = nothing
```
Comment (timholy) [inline src/GLib/signals.jl]: don't need the `local` now

**Comment by vtjnash** [comment]
This should be better now and make #161 happier

**Comment by vtjnash** [comment]
So, even though this PR doesn't fix #161, it makes it more trivial to work around:

1. if you are in a callback, don't yield. use `@async print`, etc, or queue a message for yourself, etc. You can also call `GLib.g_yield()`, but see 2

2. if you are getting segfaults in some random method due to there existing a callback that recursively calls the glib main loop (such as making a dialog box) or otherwise causes `g_yield` to be called, wrap the faulting code in `@sigatom`. This will postpone execution of the code until it can be run on the proper stack (but otherwise acts like normal control flow).

File: src/GLib/signals.jl [hunk 3]
```diff
@@ -151,165 +151,113 @@ function signal_emit(w::GObject, sig::AbstractStringLike, RT::Type, args...)
...
-gtk_work = nothing
-g_sigatom_flag = false
-function g_sigatom(f::Base.Callable)
-    # don't yield() from f inside g_sigatom, use g_siginterruptible(yield) if unsure
-    global g_sigatom_flag, gtk_stack, gtk_work
-    prev = g_sigatom_flag
-    stk = gtk_stack
+g_stack = nothing # need to call g_loop_run from only one stack
+const g_yielded = Ref(false) # when true, use the `g_doatomic` queue to run sigatom functions
+const g_doatomic = [] # (work, notification) scheduler queue
```
Comment (timholy) [inline src/GLib/signals.jl]: Is there any virtue in declaring this `Tuple{Any,Any}[]`?

**Comment by timholy** [comment]
Maybe add a version of those two comments to the README?

:100:

File: src/GLib/signals.jl (hunk 3 above)
Comment (vtjnash) [inline src/GLib/signals.jl]: I suspect that would be slower

Comment (timholy) [inline src/GLib/signals.jl]: Fine then.

**Comment by timholy** [comment]
Awesome. Tag a new release?

**Comment by vtjnash** [comment]
I figured I would give it a day or two for the bug reports to start coming in. But if it seems to be working, then sure, I can tag a new release.

**Comment by timholy** [comment]
I haven't tested it yet myself, but that may change by the end of the day.

Thanks for tackling this, I'm really looking forward to it.

**Comment by timholy** [comment]
I tested whether I could live without https://github.com/timholy/GtkUtilities.jl/pull/7 by wrapping
https://github.com/timholy/GtkUtilities.jl/blob/9ef4557c7531975ccde2f7e245596588bcfc939c/src/guidata.jl#L42-L48 in `@sigatom` and/or `@schedule`. It's much better, but I still can get the segfault.

If you need some test code, I can arrange something. That said, I know you have many priorities, so no rush.

**Comment by vtjnash** [comment]
Test code would be good. There shouldn't be any difference between guarded and siginterruptable now

**Comment by tknopp** [comment]
@vtjnash: I have tested it during last week and `sigatom` made all my segfaulting apps just work. This is very much appreciated. Would be great if a new release could be tagged.

**Pull request merged**